designs/2020-generic-ast-support/README.md

```diff
@@ -0,0 +1,57 @@
+- Repo: eslint/eslint
+- Start Date: 2020-06-04
+- RFC PR: 
+- Authors: Ilya Volodin
+
+# Generic AST support
+
+## Summary
+
+Allow ESLint to traverse and lint generic (non estree compliant AST) and allow parsers to specify information about AST that would drive ESLint's traversal and features.
+
+## Motivation
+
+ESLint has cemented itself as a go to tool for JavaScript/TypeScript ecosystem for style checking and linting. However, there are a lot of connected languages/technologies that are used every day by ESLint users that are either lacking linters, or require yet another tool to be setup and configured for the repository. Examples include HTML, CSS, SCSS, LESS, GraphQL, JSON, etc.
+With addition of config overrides users can now setup separate configuration per glob, so that would allow them to have a single instance of ESLint and a single configuration that would allow linting of the entire repository, including all non javascript/typescript files.
+
+## Detailed Design
+
+Currently ESLint has a lot of hardcoded assumptions about AST. For example, it assumes that nodes are identified by the property called `type` and that property exists on every node. It assumes that top-most node is always `Program` and that AST supports code-path-analysis (which is very specific to JavaScript). This RFC proposes to modify expected output from the `parseForESLint` method of the custom parsers to include additional property called `parserSupportOptions` of type object with four new properties (for now):
+ - nodeIdentifier: string - Specifies name of the ASTNode property that identifies node. Defaults to `type` for backwards compatibility.
+ - supportCodePathAnalysis: Boolean - Indicates AST's support for built-in CodePathAnalysis support. Defaults to `true` for backwards compatibility.
+ - supportScope: Boolean - Indicates AST's support for scope analysis. Defaults to `true` for backwards compatibility.
+ - rootNode: string - Provides the name of the top level node in the AST. Defaults to `Program` for backwards compatibility.
+
+ This would change would allow basic support of linting non-estree compliant ASTs. This will still not enable full feature parity with ESLint's abilities when linting JavaScript. In the future new properties might be added to support parsing comments to enable inline directives for controlling rules. Other enhancements might include ability to provide replacement for CodePathAnalysis and scopes creation.
```

Comment:
```suggestion
 This change would allow basic support of linting non-estree compliant ASTs. This will still not enable full feature parity with ESLint's abilities when linting JavaScript. In the future new properties might be added to support parsing comments to enable inline directives for controlling rules. Other enhancements might include ability to provide replacement for CodePathAnalysis and scopes creation.
```